File: datastructures/hashtables/java/PLHashMap.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class PLHashMap<K, V> {

    private static class Entry<K, V> {
        final K key;
        V value;

        public Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == this) return true;

            if (obj instanceof Entry) {
                Entry<K, V> entry = (Entry) obj;

                return key.equals(entry.getKey()) &&
                        value.equals(entry.getValue());
            }
            return false;
        }

        @Override
        public String toString() {
            return "{" + key + ", " + value + "}";
        }
    }

    private static class Bucket<K, V> {
        LinkedList<Entry<K, V>> bucket;

        public Bucket() {
            bucket = new LinkedList<Entry<K, V>>();
        }

        public V get(K key) {
            for (Entry<K, V> entry : bucket) {
               if(entry.key.equals(key)) {
                   return entry.value;
               } 
            }
            return null;
        }

        public void update(K key, V value) {
            boolean found = false;
            for (Entry<K, V> entry : bucket) {
                if(entry.key.equals(key)) {
                    entry.value = value;
                    found = true;
                } 
            }
            if (!found) {
                this.bucket.add(new Entry<K, V>(key, value));
            } 
        }

        public void remove(K key) {
            for (Entry<K, V> entry : bucket) {
                if(entry.key.equals(key)) {
                    this.bucket.remove(entry);
                    break;
                } 
            }
        }
    }

    private int keySpace;
    private List<Bucket<K, V>> hashTable;

    public PLHashMap() {
        this.keySpace = 2069;
        this.hashTable = new ArrayList<>();
        for (int i = 0; i < this.keySpace; i++) {
            this.hashTable.add(new Bucket<>());
        }
    }

    public void put(K key, V value) {
        int hashKey = key.hashCode() % this.keySpace;
        this.hashTable.get(hashKey).update(key, value);
    }

    public V get(K key) {
        int hashKey = key.hashCode() % this.keySpace;
        return this.hashTable.get(hashKey).get(key);
    }

    public void remove(K key) {
        int hashKey = key.hashCode() % this.keySpace;
        this.hashTable.get(hashKey).remove(key);
    }
    
    public static void main(String[] args) {
        PLHashMap<String, String> plHashMap = new PLHashMap<>();

        plHashMap.put("0", "abc");
        plHashMap.put("1", "xyz");
        plHashMap.put("test", "bla");

        System.out.println("Value at test: " + plHashMap.get("test"));
    }
}

File: datastructures/hashtables/python/pl_hashmap.py
class Bucket:
    def __init__(self):
        self.bucket = []

    def get(self, key):
        for (k, v) in self.bucket:
            if k == key:
                return v
        return None    
    
    def update(self, key, value):
        found = False
        for i, entry in enumerate(self.bucket):
            if key == entry[0]:
                self.bucket[i] = (key, value)
                found = True
                break
        if not found:
            self.bucket.append((key, value))
            
    def remove(self, key):
        for i, entry in enumerate(self.bucket):
            if key == entry[0]:
                del self.bucket[i]
                
                
class PLHashMap(object):
    def __init__(self):
        self.key_space = 2069
        self.hash_table = [Bucket() for i in range(self.key_space)]
        
    def put(self, key, value):
        hash_key = hash(key) % self.key_space
        self.hash_table[hash_key].update(key, value)
        
    def get(self, key):
        hash_key = hash(key) % self.key_space
        return self.hash_table[hash_key].get(key)
    
    def remove(self, key):
        hash_key = hash(key) % self.key_space
        self.hash_table[hash_key].remove(key)
        
        
def main():
    pl_hashmap = PLHashMap()
    
    pl_hashmap.put("A", "abc")
    pl_hashmap.put("B", "bac")
    pl_hashmap.put("C", "cab")
    
    print(f'Value for key "A" is {pl_hashmap.get("A")}')
    print(f'Value for key "B" is {pl_hashmap.get("B")}')
    print(f'Value for key "C" is {pl_hashmap.get("C")}')
    
    
if __name__ == "__main__":
    main()
            